File: src/backend/app/services/TranslationService.java
package services;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import models.TranslationConfig;
import play.db.jpa.JPAApi;
import play.libs.ws.WSClient;
import play.libs.ws.WSRequest;
import play.libs.ws.WSResponse;

import jakarta.inject.Inject;
import jakarta.inject.Singleton;
import jakarta.persistence.TypedQuery;
import java.util.concurrent.CompletionStage;
import java.util.concurrent.CompletableFuture;
import java.time.Duration;

@Singleton
public class TranslationService {

    private final JPAApi jpaApi;
    private final WSClient wsClient;
    private final ObjectMapper objectMapper;
    private static final play.Logger.ALogger logger = play.Logger.of(TranslationService.class);

    @Inject
    public TranslationService(JPAApi jpaApi, WSClient wsClient) {
        this.jpaApi = jpaApi;
        this.wsClient = wsClient;
        this.objectMapper = new ObjectMapper();
    }

    public CompletionStage<String> translateText(String text, String targetLanguage) {
        return getActiveConfiguration().thenCompose(configOpt -> {
            if (configOpt.isEmpty()) {
                logger.error("No active translation configuration found");
                return CompletableFuture.completedFuture(text); // Return original text if no config
            }

            TranslationConfig config = configOpt.get();
            return performTranslation(text, targetLanguage, config);
        });
    }

    private CompletionStage<java.util.Optional<TranslationConfig>> getActiveConfiguration() {
        return CompletableFuture.supplyAsync(() -> {
            return jpaApi.withTransaction(em -> {
                try {
                    TypedQuery<TranslationConfig> query = em.createQuery(
                        "SELECT tc FROM TranslationConfig tc WHERE tc.isActive = true ORDER BY tc.updatedAt DESC",
                        TranslationConfig.class
                    );
                    query.setMaxResults(1);
                    TranslationConfig config = query.getSingleResult();
                    return java.util.Optional.of(config);
                } catch (Exception e) {
                    logger.warn("No active translation configuration found: " + e.getMessage());
                    return java.util.Optional.<TranslationConfig>empty();
                }
            });
        });
    }

    private CompletionStage<String> performTranslation(String text, String targetLanguage, TranslationConfig config) {
        try {
            // Validate API key format
            String apiKey = config.getApiKey();
            if (apiKey == null || apiKey.trim().isEmpty()) {
                logger.error("API key is null or empty");
                return CompletableFuture.completedFuture(text);
            }
            
            // Log API key format for debugging (only first/last few characters)
            logger.info("Using API key format: {}...{} (length: {})", 
                       apiKey.substring(0, Math.min(8, apiKey.length())),
                       apiKey.length() > 8 ? apiKey.substring(apiKey.length() - 4) : "",
                       apiKey.length());

            // Create the request payload for OpenRouter API
            ObjectNode requestBody = objectMapper.createObjectNode();
            requestBody.put("model", config.getModelName());
            requestBody.put("max_tokens", config.getMaxTokens());
            requestBody.put("temperature", config.getTemperature());

            // Create messages array
            ArrayNode messages = objectMapper.createArrayNode();
            
            // System message
            ObjectNode systemMessage = objectMapper.createObjectNode();
            systemMessage.put("role", "system");
            systemMessage.put("content", buildSystemPrompt(targetLanguage));
            messages.add(systemMessage);

            // User message
            ObjectNode userMessage = objectMapper.createObjectNode();
            userMessage.put("role", "user");
            userMessage.put("content", text);
            messages.add(userMessage);

            requestBody.set("messages", messages);

            // Log request details for debugging
            logger.info("Making translation request to: {}", config.getBaseUrl() + "/chat/completions");
            logger.info("Request model: {}, max_tokens: {}, temperature: {}", 
                       config.getModelName(), config.getMaxTokens(), config.getTemperature());

            // Make the API call with proper OpenRouter headers
            WSRequest request = wsClient.url(config.getBaseUrl() + "/chat/completions")
                .addHeader("Authorization", "Bearer " + apiKey.trim())
                .addHeader("Content-Type", "application/json")
                .addHeader("HTTP-Referer", "https://secman.local")
                .addHeader("X-Title", "SecMan Translation Service")
                .setRequestTimeout(Duration.ofSeconds(30));

            return request.post(requestBody).thenApply(response -> {
                logger.info("Received response with status: {}", response.getStatus());
                logger.info("Recieved response body content: {}", response.getBody());
                return parseTranslationResponse(response);
            });

        } catch (Exception e) {
            logger.error("Error creating translation request", e);
            return CompletableFuture.completedFuture(text); // Return original text on error
        }
    }

    private String buildSystemPrompt(String targetLanguage) {
        String languageName = getLanguageName(targetLanguage);
        return String.format(
            "You are a professional translator specializing in technical documentation and risk assessment content. " +
            "Translate the following text from English to %s. " +
            "Maintain the original meaning, technical accuracy, and formal tone. " +
            "Keep technical terms consistent and preserve any formatting. " +
            "Only respond with the translated text, no explanations or additional content.",
            languageName
        );
    }

    private String getLanguageName(String languageCode) {
        switch (languageCode.toLowerCase()) {
            case "de":
            case "german":
                return "German";
            case "en":
            case "english":
                return "English";
            default:
                return languageCode;
        }
    }

    private String parseTranslationResponse(WSResponse response) {
        try {
            String responseBody = response.getBody();
            
            if (response.getStatus() != 200) {
                // Provide detailed error information for common HTTP status codes
                String errorDetail = "";
                switch (response.getStatus()) {
                    case 401:
                        errorDetail = " - Authentication failed. Check API key validity.";
                        break;
                    case 403:
                        errorDetail = " - Forbidden. API key may lack necessary permissions.";
                        break;
                    case 429:
                        errorDetail = " - Rate limit exceeded. Try again later.";
                        break;
                    case 500:
                        errorDetail = " - OpenRouter server error.";
                        break;
                    default:
                        errorDetail = " - Unknown error.";
                        break;
                }
                
                logger.error("Translation API error: HTTP {} {} - Response: {}", 
                           response.getStatus(), errorDetail, responseBody);
                
                // Try to parse error details from response body
                try {
                    JsonNode errorJson = objectMapper.readTree(responseBody);
                    if (errorJson.has("error")) {
                        JsonNode error = errorJson.get("error");
                        String errorMessage = error.has("message") ? error.get("message").asText() : "Unknown error";
                        String errorType = error.has("type") ? error.get("type").asText() : "unknown_error";
                        String errorCode = error.has("code") ? error.get("code").asText() : "unknown";
                        
                        logger.error("OpenRouter API error details - Type: {}, Code: {}, Message: {}", 
                                   errorType, errorCode, errorMessage);
                    }
                } catch (Exception parseEx) {
                    logger.warn("Could not parse error response as JSON: {}", parseEx.getMessage());
                }
                
                return ""; // Return empty string on API error
            }

            JsonNode responseJson = response.asJson();
            logger.info("Translation API response received successfully");
            
            if (responseJson.has("choices") && responseJson.get("choices").isArray() && 
                responseJson.get("choices").size() > 0) {
                
                JsonNode firstChoice = responseJson.get("choices").get(0);
                if (firstChoice.has("message") && firstChoice.get("message").has("content")) {
                    String translatedText = firstChoice.get("message").get("content").asText();
                    logger.info("Translation completed successfully. Text length: {} characters", 
                               translatedText.length());
                    return translatedText.trim();
                }
            }

            logger.error("Unexpected response format from translation API: {}", responseBody);
            return ""; // Return empty string if response format is unexpected

        } catch (Exception e) {
            logger.error("Error parsing translation response", e);
            return ""; // Return empty string on parsing error
        }
    }

    public CompletionStage<Boolean> testConfiguration(TranslationConfig config, String testText) {
        String originalText = testText != null && !testText.trim().isEmpty() ? 
                             testText : "This is a test message for translation configuration.";
        
        return performTranslation(originalText, "german", config)
            .thenApply(translatedText -> {
                boolean success = translatedText != null && !translatedText.trim().isEmpty() && 
                                 !translatedText.equals(originalText);
                
                if (success) {
                    logger.info("Translation test successful. Original: '{}', Translated: '{}'", 
                               originalText, translatedText);
                } else {
                    logger.warn("Translation test failed. Original: '{}', Result: '{}'", 
                               originalText, translatedText);
                }
                
                return success;
            })
            .exceptionally(throwable -> {
                logger.error("Translation test failed with exception", throwable);
                return false;
            });
    }
}

File: src/backendng/src/main/kotlin/com/secman/service/TranslationService.kt
package com.secman.service

import com.secman.domain.TranslationConfig
import com.secman.repository.TranslationConfigRepository
import com.fasterxml.jackson.databind.ObjectMapper
import com.github.benmanes.caffeine.cache.Cache
import com.github.benmanes.caffeine.cache.Caffeine
import io.micronaut.http.MediaType
import io.micronaut.http.uri.UriBuilder
import io.micronaut.serde.annotation.Serdeable
import jakarta.annotation.PostConstruct
import jakarta.inject.Named
import jakarta.inject.Singleton
import org.slf4j.LoggerFactory
import java.io.IOException
import java.net.URI
import java.net.http.HttpClient as JdkHttpClient
import java.net.http.HttpRequest as JdkHttpRequest
import java.net.http.HttpResponse.BodyHandlers
import java.time.Duration
import java.util.concurrent.CompletableFuture
import java.util.concurrent.ExecutorService
import java.util.concurrent.atomic.AtomicInteger

@Singleton
open class TranslationService(
    private val translationConfigRepository: TranslationConfigRepository,
    private val objectMapper: ObjectMapper,
    @Named("translation") private val translationExecutor: ExecutorService
) {

    private val logger = LoggerFactory.getLogger(TranslationService::class.java)

    private companion object {
        const val MAX_ATTEMPTS = 3
        const val INITIAL_BACKOFF_MS = 1_000L
        const val MAX_BACKOFF_MS = 8_000L
        const val BACKOFF_MULTIPLIER = 2.0
        const val CACHE_MAX_SIZE = 10_000L
        val CACHE_TTL: Duration = Duration.ofHours(24)
        val CONNECT_TIMEOUT: Duration = Duration.ofSeconds(10)
        val REQUEST_TIMEOUT: Duration = Duration.ofSeconds(120)
    }

    // In-process cache keyed by (sourceText, targetLanguage). Holds only successful
    // translations — refusals and fallbacks to the original text are NEVER cached,
    // so a transient model refusal won't pin a bad result for 24h.
    private val translationCache: Cache<Pair<String, String>, String> = Caffeine.newBuilder()
        .maximumSize(CACHE_MAX_SIZE)
        .expireAfterWrite(CACHE_TTL)
        .recordStats()
        .build()

    // JDK HttpClient — explicit, observable connection behavior. HTTP/1.1 forces one TCP
    // connection per concurrent request (no HTTP/2 stream multiplexing limit).
    //
    // CRITICAL: do NOT call .executor(translationExecutor) here. JDK HttpClient uses its
    // configured executor for response-completion tasks (reading headers, decoding body).
    // If all 8 worker threads are blocked inside .send(), and the same pool is supposed to
    // process incoming responses, the responses cannot be delivered → indefinite hang.
    // Letting JDK use its default internal cached pool decouples I/O completion from our
    // worker threads, so .send() unblocks normally.
    private val jdkHttpClient: JdkHttpClient = JdkHttpClient.newBuilder()
        .connectTimeout(CONNECT_TIMEOUT)
        .version(JdkHttpClient.Version.HTTP_1_1)
        .build()

    // Diagnostic counter — bumped at the start of each HTTP call, decremented on completion.
    // Logged with the thread name so you can verify true concurrency in production logs.
    private val inFlight = AtomicInteger(0)

    @PostConstruct
    fun logExecutorInfo() {
        logger.info(
            "TranslationService initialized: executor={}, jdkHttpClient version={}",
            translationExecutor.javaClass.name,
            JdkHttpClient.Version.HTTP_1_1
        )
    }

    @Serdeable
    data class TranslateRequest(
        val model: String,
        val messages: List<Message>,
        val max_tokens: Int? = null,
        val temperature: Double? = null
    )

    @Serdeable
    data class Message(
        val role: String,
        // Nullable: OpenRouter returns content=null for refusals, safety-filtered prompts,
        // and reasoning-only responses (e.g. o1, claude-thinking). Treated as a translation
        // miss downstream — we keep the original text rather than blowing up deserialization.
        val content: String? = null,
        val refusal: String? = null,
        val reasoning: String? = null
    )

    @Serdeable
    data class TranslateResponse(
        val id: String? = null,
        val choices: List<Choice>? = null,
        val error: ErrorDetail? = null,
        val provider: String? = null,
        val model: String? = null,
        val `object`: String? = null,
        val created: Long? = null,
        val usage: Usage? = null
    )

    @Serdeable
    data class Choice(
        val message: Message? = null,
        val finish_reason: String? = null,
        val logprobs: Any? = null,
        val native_finish_reason: String? = null,
        val index: Int? = null
    )

    @Serdeable
    data class ErrorDetail(
        val message: String? = null,
        val type: String? = null,
        val code: String? = null
    )

    @Serdeable
    data class Usage(
        val prompt_tokens: Int? = null,
        val completion_tokens: Int? = null,
        val total_tokens: Int? = null
    )

    @Serdeable
    data class TestResult(
        val success: Boolean,
        val message: String,
        val details: String? = null
    )

    /**
     * Get the active translation configuration
     */
    fun getActiveConfig(): TranslationConfig? {
        return translationConfigRepository.findActiveConfig().orElse(null)
    }

    /**
     * Translate text from English to the specified target language.
     *
     * Synchronous portion (config check + cache lookup) runs on the caller thread.
     * Actual HTTP work is dispatched to the named "translation" executor via
     * supplyAsync, which works regardless of self-invocation (unlike @Async).
     */
    open fun translateText(text: String, targetLanguage: String): CompletableFuture<String> {
        val config = getActiveConfig()
            ?: return CompletableFuture.completedFuture(text)

        if (!config.isValid()) {
            logger.warn("Translation config is not valid: {}", config)
            return CompletableFuture.completedFuture(text)
        }

        val cacheKey = text to targetLanguage
        translationCache.getIfPresent(cacheKey)?.let { cached ->
            logger.debug("Translation cache hit for '{}' [{}]", text.take(40), targetLanguage)
            return CompletableFuture.completedFuture(cached)
        }

        return CompletableFuture.supplyAsync(
            { doTranslateBlocking(text, targetLanguage, config, cacheKey) },
            translationExecutor
        )
    }

    /**
     * Blocking translation worker — runs on a translationExecutor thread.
     * Always returns a non-null String: either the translated text on success,
     * or the original text as a fallback on any failure.
     */
    private fun doTranslateBlocking(
        text: String,
        targetLanguage: String,
        config: TranslationConfig,
        cacheKey: Pair<String, String>
    ): String {
        val prompt = buildTranslationPrompt(text, targetLanguage)
        val request = TranslateRequest(
            model = config.modelName,
            messages = listOf(Message("user", prompt)),
            max_tokens = config.maxTokens,
            temperature = config.temperature
        )

        val fullUrl = "${config.baseUrl}/chat/completions"
        val requestJson = objectMapper.writeValueAsString(request)
        val jdkRequest = JdkHttpRequest.newBuilder()
            .uri(URI.create(fullUrl))
            .timeout(REQUEST_TIMEOUT)
            .header("Content-Type", MediaType.APPLICATION_JSON)
            .header("Authorization", "Bearer ${config.apiKey}")
            .header("HTTP-Referer", "https://secman.local")
            .header("X-Title", "SecMan Translation Service")
            .POST(JdkHttpRequest.BodyPublishers.ofString(requestJson))
            .build()

        return try {
            val translated = executeWithRetry("translateText[${targetLanguage}]") {
                val current = inFlight.incrementAndGet()
                logger.debug(
                    "Making translation request to: {} (in-flight={}, thread={})",
                    fullUrl, current, Thread.currentThread().name
                )
                val response = try {
                    jdkHttpClient.send(jdkRequest, BodyHandlers.ofString())
                } finally {
                    inFlight.decrementAndGet()
                }
                val statusCode = response.statusCode()
                if (statusCode in 500..599 || statusCode == 408 || statusCode == 429) {
                    throw IOException("Translation API HTTP $statusCode")
                }
                if (statusCode !in 200..299) {
                    logger.error("Translation API returned non-success status {}: {}", statusCode, response.body()?.take(200))
                    throw RuntimeException("Translation API HTTP $statusCode")
                }
                val body = response.body()
                if (body.isNullOrBlank()) {
                    throw IOException("Empty response body from translation API")
                }
                val parsed = objectMapper.readValue(body, TranslateResponse::class.java)
                if (parsed.error != null) {
                    logger.error("Translation API error: {}", parsed.error.message)
                    null
                } else {
                    val msg = parsed.choices?.firstOrNull()?.message
                    if (msg != null && msg.content.isNullOrBlank()) {
                        logger.warn(
                            "Translation produced no content (refusal='{}', reasoning='{}', finish_reason='{}') for text: {}",
                            msg.refusal?.take(120),
                            msg.reasoning?.take(120),
                            parsed.choices.firstOrNull()?.finish_reason,
                            text.take(80)
                        )
                    }
                    msg?.content
                }
            }
            if (!translated.isNullOrBlank()) {
                val trimmed = translated.trim()
                translationCache.put(cacheKey, trimmed)
                logger.info("Translation successful: '{}' -> '{}'", text.take(50), trimmed.take(50))
                trimmed
            } else {
                logger.warn("Empty translation response for text: {}", text.take(50))
                text
            }
        } catch (e: Exception) {
            logger.error("Translation failed for text after retries: {}", text.take(50), e)
            text
        }
    }

    /**
     * Run [block] with bounded exponential backoff retries on transient HTTP failures
     * (read timeouts, connection issues, 408, 429, 5xx). All other exceptions and
     * non-retryable parse-level outcomes propagate unchanged so callers can fall back.
     */
    private fun <T> executeWithRetry(operation: String, block: () -> T): T {
        var attempt = 1
        var delay = INITIAL_BACKOFF_MS
        while (true) {
            try {
                return block()
            } catch (e: Exception) {
                if (attempt >= MAX_ATTEMPTS || !isRetryable(e)) {
                    throw e
                }
                val jitter = (Math.random() * delay / 4).toLong()
                val sleep = (delay + jitter).coerceAtMost(MAX_BACKOFF_MS)
                logger.warn(
                    "{} failed (attempt {}/{}): {} — retrying in {} ms",
                    operation, attempt, MAX_ATTEMPTS, e.javaClass.simpleName, sleep
                )
                Thread.sleep(sleep)
                attempt++
                delay = (delay * BACKOFF_MULTIPLIER).toLong().coerceAtMost(MAX_BACKOFF_MS)
            }
        }
    }

    private fun isRetryable(e: Throwable): Boolean = when (e) {
        is IOException -> true            // covers JDK send() IOExceptions, our 5xx/408/429 wraps
        is InterruptedException -> false  // surface interruption immediately
        else -> false
    }

    /**
     * Translate multiple texts concurrently.
     *
     * Each translateText call submits its blocking work to translationExecutor via
     * supplyAsync, so this method fans out N futures that run in parallel up to the
     * executor's pool size (n-threads=8 in application.yml). No @Async here — this
     * method just orchestrates futures and returns immediately.
     */
    open fun translateTexts(texts: List<String>, targetLanguage: String): CompletableFuture<List<String>> {
        if (texts.isEmpty()) {
            return CompletableFuture.completedFuture(emptyList())
        }
        val futures = texts.map { translateText(it, targetLanguage) }
        return CompletableFuture.allOf(*futures.toTypedArray())
            .thenApply { _ -> futures.map { it.get() } }
    }

    /**
     * Test the translation configuration
     */
    fun testConfiguration(config: TranslationConfig): TestResult {
        if (!config.isValid()) {
            return TestResult(
                success = false,
                message = "Configuration validation failed",
                details = "Invalid API key, base URL, or model settings"
            )
        }

        return try {
            val testText = "Hello, this is a test."
            val prompt = buildTranslationPrompt(testText, "de")

            val request = TranslateRequest(
                model = config.modelName,
                messages = listOf(Message("user", prompt)),
                max_tokens = 100,
                temperature = 0.1
            )

            val fullUrl = UriBuilder.of(config.baseUrl).path("/chat/completions").build().toString()
            val jdkRequest = JdkHttpRequest.newBuilder()
                .uri(URI.create(fullUrl))
                .timeout(REQUEST_TIMEOUT)
                .header("Content-Type", MediaType.APPLICATION_JSON)
                .header("Authorization", "Bearer ${config.apiKey}")
                .header("HTTP-Referer", "https://secman.local")
                .header("X-Title", "SecMan Translation Service Test")
                .POST(JdkHttpRequest.BodyPublishers.ofString(objectMapper.writeValueAsString(request)))
                .build()

            val response = jdkHttpClient.send(jdkRequest, BodyHandlers.ofString())
            when (response.statusCode()) {
                200 -> {
                    val body = response.body()?.let { objectMapper.readValue(it, TranslateResponse::class.java) }
                    when {
                        body?.error != null -> TestResult(
                            success = false,
                            message = "API returned error: ${body.error.message}",
                            details = "Error type: ${body.error.type}, Code: ${body.error.code}"
                        )
                        body?.choices?.isNotEmpty() == true -> {
                            val translatedText = body.choices.firstOrNull()?.message?.content
                            TestResult(
                                success = true,
                                message = "Translation test successful",
                                details = "Test translation: '$testText' -> '${translatedText?.take(50)}'"
                            )
                        }
                        else -> TestResult(
                            success = false,
                            message = "Empty or invalid API response",
                            details = "Response body: ${response.body()?.take(200)}"
                        )
                    }
                }
                401 -> TestResult(success = false, message = "Authentication failed", details = "Invalid API key or insufficient permissions")
                429 -> TestResult(success = false, message = "Rate limit exceeded", details = "Too many requests to the translation API")
                else -> TestResult(success = false, message = "HTTP error ${response.statusCode()}", details = "Body: ${response.body()?.take(200)}")
            }
        } catch (e: Exception) {
            logger.error("Translation configuration test failed", e)
            TestResult(
                success = false,
                message = "Connection or request failed",
                details = e.message ?: "Unknown error"
            )
        }
    }

    /**
     * Build translation prompt for the AI model
     */
    private fun buildTranslationPrompt(text: String, targetLanguage: String): String {
        val languageName = TranslationConfig.SUPPORTED_LANGUAGES[targetLanguage] ?: targetLanguage
        
        return """
            Translate the following text from English to $languageName.
            
            Requirements:
            - Maintain the original meaning and context
            - Preserve any technical terminology appropriately
            - Keep formatting and structure intact
            - Provide only the translation without additional commentary
            
            Text to translate:
            $text
        """.trimIndent()
    }

    /**
     * Get supported languages
     */
    fun getSupportedLanguages(): Map<String, String> {
        return TranslationConfig.SUPPORTED_LANGUAGES
    }

    /**
     * Get available models
     */
    fun getAvailableModels(): Map<String, String> {
        return TranslationConfig.AVAILABLE_MODELS
    }
}